python/traverse.py

```diff
@@ -0,0 +1,179 @@
+
+    if element.tail:
+        # print("appending element tail", text)
+        stripped_element_tail = element.tail.rstrip()
+        sep = " " if stripped_element_tail and stripped_text else ""
+        element.tail = stripped_element_tail + sep + stripped_text
+    else:
+        # print("setting element tail", text)
+        element.tail = " " + stripped_text
+
+
+def traverse_and_execute(
+    html: str, fn: Callable[[lxml.html.HtmlElement], None]
+) -> list:
+    elements = lxml.html.fragments_fromstring(html)
+
+    # If there's leading text, the first element of the array will be a string.
+    # We can just discard that.
+    if isinstance(elements[0], str):
+        del elements[0]
+
+    for element in elements:
+        for e in element.iter():
+            fn(e)
+
+
+def get_source_definition(element: lxml.html.HtmlElement) -> str:
+    # https://html.spec.whatwg.org/multipage/parsing.html#escapingString
+    attributes = (
+        f'''{k}="{v.replace('"', "&quot;")}"''' for k, v in element.attrib.items()
+    )
+    return f"<{' '.join((element.tag, *attributes))}>"
+
+
+def traverse_and_replace(
+    html: str, replace: Callable[[lxml.html.HtmlElement], ElementReplacement]
+) -> str:
+    # Initialize result and work data structures
+    result: Deque[Union[str, lxml.html.HtmlElement]] = deque()
+
+    initial_list = lxml.html.fragments_fromstring(html)
+    count_stack: Deque[int] = deque([len(initial_list)])
+    work_stack: Deque[Union[str, lxml.html.HtmlElement]] = deque(reversed(initial_list))
+    tail_stack: Deque[Tuple[str, Optional[str]]] = deque()
+
+    while work_stack:
+        element = work_stack.pop()
+
+        # For just a string, append to final result
+        if isinstance(element, str):
+            result.append(element)
+
+        else:
+            new_elements = replace(element)
+
+            # Turn new_elements into a list containing we can process
+            if new_elements is None:
+                new_elements = []
+            elif isinstance(new_elements, str):
+                fragments = lxml.html.fragments_fromstring(new_elements)
+                new_elements = fragments if fragments is not None else []
+
+            if isinstance(new_elements, list):
+                # Modify count stack for new elements and decrement for element that was replaced
+                count_stack[-1] += len(new_elements) - 1
+
+                # Add element tail before processing replaced element
+                if element.tail is not None:
+                    count_stack[-1] += 1
+                    work_stack.append(element.tail)
+
+                # Extend and go to the next iteration
+                if new_elements:
+                    work_stack.extend(reversed(new_elements))
+
+                continue
+
+            if isinstance(new_elements, lxml.html.HtmlComment):
+                result.append(lxml.html.tostring(new_elements, encoding="unicode"))
+                continue
```

Comment:
I don't think there should be a `continue` here. If comments are allowed to be items in `new_elements` when it is a list, then the algorithm still needs to modify `count_stack` after appending this to the result. The continue in the above case is only used because it modifies the count stack separately. I'll admit that I was kinda going for speed when I wrote this and didn't do that great of a job cleaning up the cases 😅

Reply:
If we don't `continue` here, we try to do a bunch of things that don't make sense for comments (access its `.tag`, count its children, etc). With enough conditionals after this, we *could* forgo the `continue`, but IMO it's much easier to understand this if we just special-case comments and immediately stringify them. But maybe I'm misunderstanding the point about `count_stack` - can you write a test case that would fail with the current code?

Reply:
I think immediately stringifying is the way to go, if we have a comment, but we just need to wrap that in an if-else case with the stuff that comes after this. The following test case fails (and this should fail on any test case where you have to close a tag after inserting a comment):
```python

def test_traverse_and_replace_comment() -> None:
    def replace(e) -> ElementReplacement:
        if e.tag == "p":
            return "<!-- Goodbye --><i>world</i>"
        return e

    html = traverse_and_replace("<i><p>Hello</p></i>", replace)
    assert html == "<i><!-- Goodbye --><i>world</i></i>"
```

Reply:
I found another test case that didn't work and committed both it and your to the repo. I tried a few variations of `if`/`else` here, but it wasn't immediately obvious to me how to get the tests passing. If you understand the fix, feel free to add it as a suggestion!

Reply:
> Ran into another error trying to trace through this test case, it seems like the fragments_fromstring function sometimes doesn't return the top-level list as I would have expected? Very strange.

Can you elaborate?

Reply:
Sounds good, I'll play with it more a little later. There's another bug that comes out of some weird behavior from lxml, but I think the comment fix should be straightforward.

Reply:
@eliotwrobson I don't actually remember what the resolution here was. Did this bug become a test case? Is it fixed?